pkg/deploy/registry/instantiate/rest.go

```diff
@@ -0,0 +1,286 @@
+		switch causes[0].Type {
+		case deployapi.DeploymentTriggerOnConfigChange:
+			config.Status.Details.Message = "caused by a config change"
+		case deployapi.DeploymentTriggerOnImageChange:
+			config.Status.Details.Message = "caused by an image change"
+		case deployapi.DeploymentTriggerManual:
+			config.Status.Details.Message = "caused by a manual change"
+		}
+		config.Status.LatestVersion++
+	}
+
+	// If we cannot trigger and no image was updated then there is nothing to do here.
+	if !canTrigger && !updated {
+		return config, false, nil
+	}
+
+	return r.store.Update(ctx, config.Name, rest.DefaultUpdatedObjectInfo(config, kapi.Scheme))
+}
+
+// Process will go over all deployment triggers that require processing and update the
+// deployment config accordingly. This contains the work that the image change controller
+// had been doing up to the point we got the /instantiate endpoint.
+func (r *REST) Process(config *deployapi.DeploymentConfig, force bool) (bool, error) {
+	errs := []error{}
+	updated := false
+
+	// Process any image change triggers.
+	for _, trigger := range config.Spec.Triggers {
+		if trigger.Type != deployapi.DeploymentTriggerOnImageChange {
+			continue
+		}
+
+		params := trigger.ImageChangeParams
+
+		// Forced deployments should always try to resolve the images in the template.
+		// On the other hand, paused deployments or non-automatic triggers shouldn't.
+		if !force && (config.Spec.Paused || !params.Automatic) && len(params.LastTriggeredImage) > 0 {
+			continue
+		}
+
+		// Tag references are already validated
+		name, tag, _ := imageapi.SplitImageStreamTag(params.From.Name)
+		stream, err := r.isn.ImageStreams(params.From.Namespace).Get(name)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				errs = append(errs, err)
+			}
+			continue
+		}
+
+		// Find the latest tag event for the trigger reference.
+		latestEvent := imageapi.LatestTaggedImage(stream, tag)
+		if latestEvent == nil {
+			continue
+		}
+
+		// Ensure a change occurred
+		latestRef := latestEvent.DockerImageReference
+		if len(latestRef) == 0 || latestRef == params.LastTriggeredImage {
+			continue
+		}
+
+		// Update containers
+		names := sets.NewString(params.ContainerNames...)
+		for i := range config.Spec.Template.Spec.Containers {
+			container := &config.Spec.Template.Spec.Containers[i]
+			if !names.Has(container.Name) {
+				continue
+			}
+
+			if container.Image != latestRef {
+				// Update the image
+				container.Image = latestRef
+				// Log the last triggered image ID
+				params.LastTriggeredImage = latestRef
+				updated = true
+			}
+		}
+	}
+
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		return updated, errors.NewInternalError(err)
+	}
+
+	return updated, nil
+}
+
+// CanTrigger determines if we can trigger a new deployment for config based on the various deployment triggers.
+func (r *REST) CanTrigger(config *deployapi.DeploymentConfig, force bool) (bool, []deployapi.DeploymentCause, error) {
```

Comment:
this should also not be a method on rest, and should be private and take arguments.